File: atcoder/tyama_atcodertenka1~2012qualbB.rb
#!/usr/bin/ruby

def snake2camel(orig)
    r=''.dup
    i=0
    while i<orig.size
        if orig[i]=='_'
            return orig if i+1==orig.size || orig[i+1]=='_' || ('0'<=orig[i+1]&&orig[i+1]<='9')
            r<<orig[i+1].upcase
            i+=1
        else
            r<<orig[i]
        end
        i+=1
    end
    return r
end

def camel2snake(orig)
    r=''.dup
    orig.chars{|c|
        if 'A'<=c&&c<='Z'
            r<<'_'
            r<<c.downcase
        else
            r<<c
        end
    }
    return r
end

def convert(orig)
    s=orig
    start_underscore=0
    end_underscore=0
    while s.size>0 && s[0]=='_'
        start_underscore+=1
        s=s[1..-1]
    end
    return orig if s.empty?
    return orig if ('A'<=s[0]&&s[0]<='Z') || ('0'<=s[0]&&s[0]<='9')
    while s[-1]=='_'
        end_underscore+=1
        s=s[0...-1]
    end
    if s.index('_')
        return orig if s.chars.any?{|c|'A'<=c&&c<='Z'}
        return '_'*start_underscore+snake2camel(s)+'_'*end_underscore
    else
        return '_'*start_underscore+camel2snake(s)+'_'*end_underscore
    end
end

$<.each{|e|puts convert e.chomp}

File: atcoder/tyama_atcodertenka1~2012qualbB.py
#!/usr/bin/python
from __future__ import print_function
import sys,re

def snake2camel(orig):
    r=[]
    i=0
    while i<len(orig):
        if orig[i]=='_':
            if i+1==len(orig) or orig[i+1]=='_' or '0'<=orig[i+1]<='9': return orig
            r.append(orig[i+1].upper())
            i+=1
        else:
            r.append(orig[i])
        i+=1
    return ''.join(r)

def camel2snake(orig):
    r=[]
    for c in orig:
        if 'A'<=c<='Z':
            r.append('_')
            r.append(c.lower())
        else:
            r.append(c)
    return ''.join(r)

def convert(orig):
    s=orig
    start_underscore=0
    end_underscore=0
    while len(s)>0 and s[0]=='_':
        start_underscore+=1
        s=s[1:]
    if not s: return orig
    if 'A'<=s[0]<='Z' or '0'<=s[0]<='9': return orig
    while s[-1]=='_':
        end_underscore+=1
        s=s[:-1]
    if '_' in s:
        if any('A'<=c<='Z' for c in s): return orig
        return '_'*start_underscore+snake2camel(s)+'_'*end_underscore
    else:
        return '_'*start_underscore+camel2snake(s)+'_'*end_underscore

for line in sys.stdin:
    print(convert(line.rstrip()))
